storage_image_product/models/product_product.py

```diff
@@ -8,24 +10,22 @@
 class ProductProduct(models.Model):
     _inherit = "product.product"
 
-    # small and medium image are here to replace
-    # native image field on form and kanban
-    variant_image_small_url = fields.Char(
-        compute="_compute_variant_image_ids",
-        store=True,
-        string="Variant Image Small Url",
-    )
-    variant_image_medium_url = fields.Char(
-        compute="_compute_variant_image_ids",
-        store=True,
-        string="Variant Image Medium Url",
-    )
     variant_image_ids = fields.Many2many(
         "product.image.relation",
         compute="_compute_variant_image_ids",
         store=True,
         string="Variant Images",
     )
+    main_image_id = fields.Many2one(
+        "storage.image",
+        compute="_compute_main_image_id",
+        # Store it to improve perf on product views
+        store=True,
+    )
+    # small and medium image are here to replace
+    # native image field on form and kanban
+    variant_image_small_url = fields.Char(related="main_image_id.image_small_url")
```

Comment:
@simahawk IMO it's important to keep the image urls stored on the product template and on the variant to avoid extra queries when the product kanban view is displayed. 
If you change your env, you are not sure at all that your images exist into the other env. In this case, the new computed url could not exist.

Reply:
@lmignon but THIS is the problem. If you change environment you should not see the images IF they are not there. Otherwise, what happens is that you expect to have those images in the system but you don't. Then the customer or a functional colleague comes to you and asks "hey, I see my images in Odoo, why I don't see them in the shop (or any other service)?".

Reply:
also. if you need to change CDN URL on production, it won't have any effect on those URLs. That sucks IMO. One of the key point about storage files is that their URL is controlled by the backend.

Reply:
> @lmignon but THIS is the problem. If you change environment you should not see the images IF they are not there. Otherwise, what happens is that you expect to have those images in the system but you don't. Then the customer or a functional colleague comes to you and asks "hey, I see my images in Odoo, why I don't see them in the shop (or any other service)?".

If you change environment and you don't want to see the image, you should apply a script to remove   `product.image.relation` and related `storage.image`and to set urls to blank.
Sometime it's nice to be able to see the images from an other env. (ex test env can see image from prod env but is not allowed to modify these items.)

Reply:
> also. if you need to change CDN URL on production, it won't have any effect on those URLs. That sucks IMO. One of the key point about storage files is that their URL is controlled by the backend.

I prefer to write a script to upgrade urls by hand than fighting against performance issue... :smirk:

We could also add a listener on the write of `storage.backend` to update image urls if the URL changes on the backend in a efficient way (SQL ...)

Reply:
> Sometime it's nice to be able to see the images from an other env. (ex test env can see image from prod env but is not allowed to modify these items.)

That's another kind of problem. We can work on it. Also, setting the production CDN on other envs won't have any impact on what you can write -> easy to do, fast result, no script / black magic needed.

> I prefer to write a script to upgrade urls by hand than fighting against performance issue... smirk

I think you are not considering the changes her: very likely perf were bad because of full awakening on several m2m fields: 

* rel w/ images
* rel w/ attribute values from the image relation
* rel w/ product attribute values on variants

a lot of m2m loaded every single time which we know to be very slow!
With this change you have only a m2o to look for.

Furthermore, the URL before my refactoring on https://github.com/OCA/storage/pull/88 was fully computed. The most expensive thing being the slug, which is now stored.

And now that I think about this again I think I'm going to store the main thumbs relation as well because to compute the thunb URL we always loop on all the thumbs and it can be expensive.

Reply:
@lmignon so, in [this commit](https://github.com/OCA/storage/pull/94/commits/596dcf7a73fd4da9fd4798c4ef964a0f1350e4af)  `fixup! storage_thumb: do not store url` I've made the main thumbs computed. This ensures that the URL computation is very fast. I've tried on a kanban w/ thousands of records.

I've done a test by playing around w/ storage backend server env params and - as expected - if you keep the CDN from production (or whatever env the db is coming from) you'll see the images w/out any problem.

The only issue that I've found is that `base_url_for_files` is not refreshed when changing server env params as the change is not detected. Hence, I added a btn to refresh the base URL on the fly (this affects only the backend, no storage file computation is involved).

Sounds good?